building/build-debs/homeworld-keysystem/src/keycommon/endpoint/endpoint.go

```diff
@@ -0,0 +1,108 @@
+package endpoint
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+type ServerEndpoint struct {
+	rootCAs      *x509.CertPool
+	baseURL      string
+	extraHeaders map[string]string
+	certificates []tls.Certificate
+	timeout      time.Duration
+}
+
+func NewServerEndpoint(url string, authorities *x509.CertPool) (ServerEndpoint, error) {
+	if url == "" {
+		return ServerEndpoint{}, errors.New("empty base URL")
+	}
+	if !strings.HasSuffix(url, "/") {
+		return ServerEndpoint{}, errors.New("base URL must end in a slash")
+	}
+	return ServerEndpoint{rootCAs: authorities, baseURL: url, timeout: time.Second * 30}, nil
+}
+
+func (s ServerEndpoint) BaseURL() string {
+	return s.baseURL
+}
+
+func (s ServerEndpoint) WithHeader(key string, value string) ServerEndpoint {
+	out := s
+	out.extraHeaders = map[string]string{}
+	for k, v := range s.extraHeaders {
+		out.extraHeaders[k] = v
+	}
+	out.extraHeaders[key] = value
+	return out
+}
+
+func (s ServerEndpoint) WithCertificate(cert tls.Certificate) ServerEndpoint {
+	out := s
+	out.certificates = make([]tls.Certificate, len(s.certificates))
+	copy(out.certificates, s.certificates)
+	out.certificates = append(out.certificates, cert)
+	return out
+}
```

Comment:
I think this entire function body could be replaced with

    s.certificates = append(s.certificates, cert)
    return s

(Unless it is necessary that the new `certificates` slice does not share space with the old one.  But I don't see any reason why that should be so.)

Reply:
It is necessary -- you could call two different `WithCertificate` calls on the same endpoint, and they should produce endpoints that don't conflict.

Reply:
Oh, I see.  (Still no need for `out`.)